File: apps/server/src/routes/webhooks.ts
// Webhook subscription routes — POST/GET/DELETE /v1/webhooks
// + GET /v1/webhooks/:id/deliveries.

import type { FastifyInstance, FastifyRequest } from 'fastify';
import { knownRequestKeys, reportUnknownRequestFields } from '../lib/unknown-request-fields.js';
import {
  CreateWebhookRequestSchema,
  ListDeliveriesQuerySchema,
  UpdateWebhookRequestSchema,
} from '@driftstack/api-types';
import { BadRequestError, ForbiddenError } from '../lib/errors.js';
import type {
  WebhookDeliveryRow,
  WebhookEndpointRow,
  WebhooksService,
} from '../services/webhooks.js';
import { resolveEffectiveAccount } from '../services/auth.js';
import { readEffectiveAccountHeader } from '../lib/effective-account-header.js';
import { unsafeWebhookTargetReason } from '../lib/webhook-target-guard.js';

/**
 * V-326e5 — admin-only gate for webhook write operations on team
 * owners. Returns the effective accountId (string) when team write
 * should proceed, or undefined when self-scoped. Throws ForbiddenError
 * on member-role team requests.
 */
function effectiveAccountIdForWrite(
  request: FastifyRequest,
  ctx: NonNullable<FastifyRequest['account']>,
): string | undefined {
  const effective = resolveEffectiveAccount(ctx, readEffectiveAccountHeader(request));
  if (effective.kind !== 'team') return undefined;
  if (effective.role !== 'admin') {
    throw new ForbiddenError('Webhook writes on a team owner require admin role on that team.');
  }
  return effective.accountId;
}

const PUBLIC_ID_RE = /^[a-z]{3}_([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})$/;

function uuidFromPrefixedId(value: string, expectedPrefix: string): string {
  const match = PUBLIC_ID_RE.exec(value);
  if (!match || !match[1] || !value.startsWith(`${expectedPrefix}_`)) {
    throw new BadRequestError(`Invalid id format. Expected "${expectedPrefix}_<uuid>".`);
  }
  return match[1];
}

function publicEndpoint(
  row: WebhookEndpointRow,
  counts: { delivered: number; failed: number; dlq: number } = {
    delivered: 0,
    failed: 0,
    dlq: 0,
  },
): Record<string, unknown> {
  // V-359 — surface the rotation grace state when active. The previous
  // secret's first-12-chars are non-sensitive (same shape as the
  // current secret_prefix display); the grace expiry lets the
  // dashboard show "rotation ends in <X>" so customers know how much
  // longer dual-signing is in effect. Both null when no rotation in
  // flight.
  const rotationActive =
    row.secretPrev !== null &&
    row.secretPrevExpiresAt !== null &&
    row.secretPrevExpiresAt.getTime() > Date.now();
  return {
    id: `whk_${row.id}`,
    url: row.url,
    secret_prefix: row.secretPrefix,
    prev_secret_prefix:
      rotationActive && row.secretPrev !== null ? row.secretPrev.slice(0, 12) : null,
    rotation_grace_expires_at:
      rotationActive && row.secretPrevExpiresAt !== null
        ? row.secretPrevExpiresAt.toISOString()
        : null,
    events: row.events,
    description: row.description,
    active: row.active,
    consecutive_failures: row.consecutiveFailures,
    last_success_at: row.lastSuccessAt ? row.lastSuccessAt.toISOString() : null,
    last_failure_at: row.lastFailureAt ? row.lastFailureAt.toISOString() : null,
    disabled_at: row.disabledAt ? row.disabledAt.toISOString() : null,
    /** V-185 — aggregate per-endpoint delivery counts. */
    delivery_counts: counts,
    created_at: row.createdAt.toISOString(),
  };
}

function publicDelivery(row: WebhookDeliveryRow): Record<string, unknown> {
  return {
    id: `wdl_${row.id}`,
    webhook_id: `whk_${row.webhookId}`,
    event_id: row.eventId,
    event_type: row.eventType,
    status: row.status,
    attempts: row.attempts,
    next_attempt_at: row.nextAttemptAt.toISOString(),
    last_response_status: row.lastResponseStatus,
    last_response_excerpt: row.lastResponseExcerpt,
    last_error: row.lastError,
    delivered_at: row.deliveredAt ? row.deliveredAt.toISOString() : null,
    created_at: row.createdAt.toISOString(),
  };
}

export interface WebhookRoutesOptions {
  service: WebhooksService;
}

export function registerWebhookRoutes(app: FastifyInstance, opts: WebhookRoutesOptions): void {
  const { service } = opts;

  // V-326e5 — admin-only when targeting a team owner.
  app.post(
    '/v1/webhooks',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request, reply) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const rawBody = request.body ?? {};
      const body = CreateWebhookRequestSchema.parse(rawBody);
      // Item 6 — `events` is the field that matters here. A mistyped key is
      // stripped, the endpoint is created subscribed to whatever the schema
      // defaults to, and the customer sees 201 followed by deliveries that never
      // arrive for the events they thought they had chosen.
      reportUnknownRequestFields({
        body: rawBody,
        knownKeys: Object.keys(CreateWebhookRequestSchema.shape),
        reply,
        logger: request.log,
        route: 'POST /v1/webhooks',
      });
      // SSRF guard — reject a webhook URL pointed at a private/loopback/
      // reserved address or localhost (the delivery worker runs on our infra).
      const unsafe = unsafeWebhookTargetReason(body.url);
      if (unsafe !== null) throw new BadRequestError(unsafe);
      const eff = effectiveAccountIdForWrite(request, ctx);
      const created = await service.create(
        ctx,
        {
          url: body.url,
          events: body.events,
          description: body.description ?? null,
        },
        eff !== undefined ? { effectiveAccountId: eff } : {},
      );
      return reply.code(201).send({
        ...publicEndpoint(created.row),
        secret: created.plaintextSecret,
      });
    },
  );

  // V-330f — read endpoints + per-endpoint counts, scoped to the
  // OWNER when X-Driftstack-Account is set. Read-only; both roles
  // allowed.
  //
  // V-804 — this used to add that the write verbs stayed self-scoped
  // pending the V-326e write-side cycle. That cycle landed: POST,
  // DELETE, PATCH, rotate-secret, test and delivery-replay all resolve
  // the target through `effectiveAccountIdForWrite`, which is the V-326e5
  // admin-only gate — a member-role team request gets a ForbiddenError,
  // an admin acts on the owner. Reads allow both roles; writes require
  // admin. Nothing on this route is self-only any more.
  app.get(
    '/v1/webhooks',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const effective = resolveEffectiveAccount(ctx, readEffectiveAccountHeader(request));
      const rowsWithCounts = await service.listWithCounts(
        ctx,
        effective.kind === 'team' ? { effectiveAccountId: effective.accountId } : {},
      );
      return { data: rowsWithCounts.map((r) => publicEndpoint(r.endpoint, r.counts)) };
    },
  );

  app.get<{ Params: { id: string } }>(
    '/v1/webhooks/:id',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const effective = resolveEffectiveAccount(ctx, readEffectiveAccountHeader(request));
      const id = uuidFromPrefixedId(request.params.id, 'whk');
      // V-1603 — the counts are published on this response too, and passing no
      // second argument here returned the zero default while the list route
      // returned real numbers for the same endpoint.
      const { endpoint, counts } = await service.getWithCounts(
        ctx,
        id,
        effective.kind === 'team' ? { effectiveAccountId: effective.accountId } : {},
      );
      return publicEndpoint(endpoint, counts);
    },
  );

  // V-326e5 — admin-only when targeting a team owner.
  app.delete<{ Params: { id: string } }>(
    '/v1/webhooks/:id',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request, reply) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const id = uuidFromPrefixedId(request.params.id, 'whk');
      const eff = effectiveAccountIdForWrite(request, ctx);
      await service.delete(ctx, id, eff !== undefined ? { effectiveAccountId: eff } : {});
      return reply.code(204).send();
    },
  );

  // V-351 — partial update. Mirror of POST + DELETE for the
  // V-326e5 admin-only-on-team gate. Disabled endpoints cannot be
  // updated (the repo enforces; this returns 409).
  app.patch<{ Params: { id: string } }>(
    '/v1/webhooks/:id',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request, reply) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const id = uuidFromPrefixedId(request.params.id, 'whk');
      const parsed = UpdateWebhookRequestSchema.safeParse(request.body);
      if (!parsed.success)
        throw new BadRequestError(parsed.error.issues[0]?.message ?? 'Invalid body.');
      // A partial update makes this worse than create: every field is optional,
      // so a mistyped `events` leaves the existing subscriptions untouched and
      // reports success, and the customer believes they changed something.
      reportUnknownRequestFields({
        body: request.body,
        knownKeys: knownRequestKeys(UpdateWebhookRequestSchema),
        reply,
        logger: request.log,
        route: 'PATCH /v1/webhooks/:id',
      });
      // SSRF guard on a changed URL (partial update; url is optional).
      if (parsed.data.url !== undefined) {
        const unsafe = unsafeWebhookTargetReason(parsed.data.url);
        if (unsafe !== null) throw new BadRequestError(unsafe);
      }
      const eff = effectiveAccountIdForWrite(request, ctx);
      const row = await service.update(
        ctx,
        id,
        parsed.data,
        eff !== undefined ? { effectiveAccountId: eff } : {},
      );
      // V-1603 — same publication, same omission: an update answered with zeroed
      // counts while the list showed the endpoint's real delivery history.
      const { counts } = await service.getWithCounts(
        ctx,
        id,
        eff !== undefined ? { effectiveAccountId: eff } : {},
      );
      return publicEndpoint(row, counts);
    },
  );

  app.get<{ Params: { id: string } }>(
    '/v1/webhooks/:id/deliveries',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const id = uuidFromPrefixedId(request.params.id, 'whk');
      const query = ListDeliveriesQuerySchema.parse(request.query ?? {});
      const effective = resolveEffectiveAccount(ctx, readEffectiveAccountHeader(request));
      const page = await service.listDeliveries(ctx, id, {
        limit: query.limit,
        ...(query.cursor !== undefined ? { cursor: query.cursor } : {}),
        ...(query.status !== undefined ? { status: query.status } : {}),
        ...(effective.kind === 'team' ? { effectiveAccountId: effective.accountId } : {}),
      });
      return {
        data: page.items.map(publicDelivery),
        has_more: page.nextCursor !== null,
        next_cursor: page.nextCursor,
      };
    },
  );

  // V-307 — customer self-service replay. Different from the admin
  // /v1/admin/webhook-deliveries/:id/replay (which can replay any
  // account's delivery): this one is scoped to the EFFECTIVE account and
  // 404s otherwise — your own, or the owner you are acting as via
  // X-Driftstack-Account.
  //
  // V-1122 — this read "the calling account". The gate below resolves
  // effectiveAccountIdForWrite, so a team ADMIN replays the owner's
  // delivery; the service comment on replayDeliveryAsCustomer has said so
  // since S32 and this one was left behind.
  app.post<{ Params: { deliveryId: string } }>(
    '/v1/webhook-deliveries/:deliveryId/replay',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request, reply) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const deliveryId = uuidFromPrefixedId(request.params.deliveryId, 'wdl');
      // S32 2026-07-07 honoured team act-as here, but replay RE-FIRES the
      // delivery — a write — so it takes the same admin-only-on-team gate as
      // create/update/delete/rotate (effectiveAccountIdForWrite throws for a
      // member role), NOT the read-only act-as of listDeliveries. Without it a
      // non-admin team member could replay the owner's deliveries. (Fable
      // audit-2 2026-07-08, C5.)
      const eff = effectiveAccountIdForWrite(request, ctx);
      const updated = await service.replayDeliveryAsCustomer(ctx, deliveryId, {
        ...(eff !== undefined ? { effectiveAccountId: eff } : {}),
      });
      return reply.code(200).send(publicDelivery(updated));
    },
  );

  // V-359 — rotate the signing secret with a 24h grace. New plaintext
  // returned ONCE; worker dual-signs every outbound delivery with both
  // the new + old secret while `secret_prev_expires_at > now`.
  // Admin-only on team-scoped requests (same gate as create / update /
  // delete / send-test).
  app.post<{ Params: { id: string } }>(
    '/v1/webhooks/:id/rotate-secret',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request, reply) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const id = uuidFromPrefixedId(request.params.id, 'whk');
      const eff = effectiveAccountIdForWrite(request, ctx);
      const result = await service.rotateSecret(
        ctx,
        id,
        eff !== undefined ? { effectiveAccountId: eff } : {},
      );
      return reply.code(200).send({
        id: `whk_${result.row.id}`,
        secret: result.plaintextSecret,
        secret_prefix: result.row.secretPrefix,
        prev_secret_prefix: result.row.secretPrev ? result.row.secretPrev.slice(0, 12) : '',
        grace_expires_at: result.row.secretPrevExpiresAt
          ? result.row.secretPrevExpiresAt.toISOString()
          : new Date().toISOString(),
      });
    },
  );

  // V-356 — send a synthetic test.ping event to the endpoint,
  // bypassing subscription. Lets the customer verify their handler
  // before relying on it for real events. Admin-only when targeting
  // a team owner (same gate as create / update / delete).
  app.post<{ Params: { id: string } }>(
    '/v1/webhooks/:id/test',
    { preHandler: [app.requireAuth, app.rateLimit('global')] },
    async (request, reply) => {
      const ctx = request.account;
      if (!ctx) throw new Error('account context missing after requireAuth');
      const id = uuidFromPrefixedId(request.params.id, 'whk');
      const eff = effectiveAccountIdForWrite(request, ctx);
      const result = await service.sendTestEvent(
        ctx,
        id,
        eff !== undefined ? { effectiveAccountId: eff } : {},
      );
      return reply.code(202).send({
        delivery_id: `wdl_${result.deliveryId}`,
        event_id: result.eventId,
        event_type: 'test.ping',
      });
    },
  );
}

File: packages/sdk-python/src/driftstack/resources/webhooks.py
"""Webhooks resource — /v1/webhooks."""

from __future__ import annotations

from collections.abc import AsyncIterator, Iterator
from typing import Any
from urllib.parse import quote

from pydantic import BaseModel

from driftstack._generated.models import (
    CreateWebhookRequest,
    CreateWebhookResponse,
    ListDeliveriesQuery,
    WebhookDelivery,
    WebhookEndpoint,
)
from driftstack.http import AsyncHttpClient, HttpClient, parse_model
from driftstack.pagination import aiterate_paginated, iterate_paginated
from driftstack.resources._common import coerce_body, coerce_query


class WebhookEndpointList(BaseModel):
    """Response shape for ``GET /v1/webhooks``."""

    data: list[WebhookEndpoint]


class WebhookDeliveryListPage(BaseModel):
    """Response shape for ``GET /v1/webhooks/{id}/deliveries``."""

    data: list[WebhookDelivery]
    has_more: bool
    next_cursor: str | None


def _webhook_path(webhook_id: str, suffix: str = "") -> str:
    return f"/v1/webhooks/{quote(webhook_id, safe='')}{suffix}"


class WebhooksResource:
    """Synchronous webhooks resource."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(self, body: CreateWebhookRequest | dict[str, Any]) -> CreateWebhookResponse:
        """Create a webhook subscription.

        Plaintext signing secret is returned ONCE; store it now — it
        cannot be retrieved later. Requires the ``account_owner`` scope.
        """
        data = self._http.request("POST", "/v1/webhooks", json_body=coerce_body(body))
        return parse_model(CreateWebhookResponse, data)

    def list(self) -> WebhookEndpointList:
        data = self._http.request("GET", "/v1/webhooks")
        return parse_model(WebhookEndpointList, data)

    def get(self, webhook_id: str) -> WebhookEndpoint:
        data = self._http.request("GET", _webhook_path(webhook_id))
        return parse_model(WebhookEndpoint, data)

    def delete(self, webhook_id: str) -> None:
        """Soft-delete (disable) the endpoint. Idempotent."""
        self._http.request("DELETE", _webhook_path(webhook_id))

    def list_deliveries(
        self,
        webhook_id: str,
        query: ListDeliveriesQuery | dict[str, Any] | None = None,
    ) -> WebhookDeliveryListPage:
        data = self._http.request(
            "GET",
            _webhook_path(webhook_id, "/deliveries"),
            params=coerce_query(query),
        )
        return parse_model(WebhookDeliveryListPage, data)

    def iterate_deliveries(
        self,
        webhook_id: str,
        *,
        limit: int | None = None,
        status: str | None = None,
    ) -> Iterator[WebhookDelivery]:
        """Lazily walk every delivery for an endpoint.

        Filter by ``status`` (e.g. ``'dlq'``) to walk just one bucket;
        the filter threads through every page.
        """

        def fetch_page(cursor: str | None) -> WebhookDeliveryListPage:
            params: dict[str, Any] = {}
            if limit is not None:
                params["limit"] = limit
            if status is not None:
                params["status"] = status
            if cursor is not None:
                params["cursor"] = cursor
            return self.list_deliveries(webhook_id, params)

        return iterate_paginated(fetch_page)

    def replay_delivery(self, delivery_id: str) -> WebhookDelivery:
        """V-307 — replay a webhook delivery.

        Resets the delivery to ``pending`` so the worker re-fires it.
        Account-scoped: the delivery must belong to an endpoint the
        calling account owns.
        """
        data = self._http.request(
            "POST",
            f"/v1/webhook-deliveries/{quote(delivery_id, safe='')}/replay",
            json_body={},
        )
        return parse_model(WebhookDelivery, data)

    def rotate_secret(self, webhook_id: str) -> dict[str, Any]:
        """V-359 — rotate the webhook signing secret.

        Returns the fresh plaintext (shown ONCE) plus grace metadata:
        the previous secret stays active for 24h
        (``grace_expires_at``) during which Driftstack dual-signs
        every outbound delivery (both new + old HMAC). Roll the new
        secret across your verifier infra inside that window.
        """
        return self._http.request("POST", _webhook_path(webhook_id, "/rotate-secret"), json_body={})

    def send_test(self, webhook_id: str) -> dict[str, Any]:
        """V-356 — send a synthetic ``test.ping`` event to the endpoint.

        Bypasses subscription so customers can verify their handler
        is reachable + signature-valid before depending on it for
        real events. Returns ``{delivery_id, event_id, event_type}``.
        """
        return self._http.request("POST", _webhook_path(webhook_id, "/test"), json_body={})

    def update(self, webhook_id: str, body: dict[str, Any]) -> WebhookEndpoint:
        """V-351 — partial-update a webhook endpoint.

        At least one of ``url``, ``events``, ``description``, or
        ``active`` must be present. The signing secret is NOT rotated
        by update; use :meth:`rotate_secret` for that. Disabled
        endpoints cannot be updated (returns 409).
        """
        data = self._http.request("PATCH", _webhook_path(webhook_id), json_body=coerce_body(body))
        return parse_model(WebhookEndpoint, data)


class AsyncWebhooksResource:
    """Async webhooks resource."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def create(self, body: CreateWebhookRequest | dict[str, Any]) -> CreateWebhookResponse:
        data = await self._http.request("POST", "/v1/webhooks", json_body=coerce_body(body))
        return parse_model(CreateWebhookResponse, data)

    async def list(self) -> WebhookEndpointList:
        data = await self._http.request("GET", "/v1/webhooks")
        return parse_model(WebhookEndpointList, data)

    async def get(self, webhook_id: str) -> WebhookEndpoint:
        data = await self._http.request("GET", _webhook_path(webhook_id))
        return parse_model(WebhookEndpoint, data)

    async def delete(self, webhook_id: str) -> None:
        await self._http.request("DELETE", _webhook_path(webhook_id))

    async def list_deliveries(
        self,
        webhook_id: str,
        query: ListDeliveriesQuery | dict[str, Any] | None = None,
    ) -> WebhookDeliveryListPage:
        data = await self._http.request(
            "GET",
            _webhook_path(webhook_id, "/deliveries"),
            params=coerce_query(query),
        )
        return parse_model(WebhookDeliveryListPage, data)

    def iterate_deliveries(
        self,
        webhook_id: str,
        *,
        limit: int | None = None,
        status: str | None = None,
    ) -> AsyncIterator[WebhookDelivery]:
        """Async variant of :meth:`WebhooksResource.iterate_deliveries`."""

        async def fetch_page(cursor: str | None) -> WebhookDeliveryListPage:
            params: dict[str, Any] = {}
            if limit is not None:
                params["limit"] = limit
            if status is not None:
                params["status"] = status
            if cursor is not None:
                params["cursor"] = cursor
            return await self.list_deliveries(webhook_id, params)

        return aiterate_paginated(fetch_page)

    async def replay_delivery(self, delivery_id: str) -> WebhookDelivery:
        """V-307 — async replay. See :meth:`WebhooksResource.replay_delivery`."""
        data = await self._http.request(
            "POST",
            f"/v1/webhook-deliveries/{quote(delivery_id, safe='')}/replay",
            json_body={},
        )
        return parse_model(WebhookDelivery, data)

    async def rotate_secret(self, webhook_id: str) -> dict[str, Any]:
        """V-359 — async secret rotation. See :meth:`WebhooksResource.rotate_secret`."""
        return await self._http.request(
            "POST", _webhook_path(webhook_id, "/rotate-secret"), json_body={}
        )

    async def send_test(self, webhook_id: str) -> dict[str, Any]:
        """V-356 — async test ping. See :meth:`WebhooksResource.send_test`."""
        return await self._http.request("POST", _webhook_path(webhook_id, "/test"), json_body={})

    async def update(self, webhook_id: str, body: dict[str, Any]) -> WebhookEndpoint:
        """V-351 — async partial-update. See :meth:`WebhooksResource.update`."""
        data = await self._http.request(
            "PATCH", _webhook_path(webhook_id), json_body=coerce_body(body)
        )
        return parse_model(WebhookEndpoint, data)
